src/activity_handlers.cpp

```diff
@@ -1486,18 +1482,26 @@ void activity_handlers::forage_finish( player_activity *act, player *p )
         default:
             debugmsg( "Invalid season" );
     }
+    return next_ter;
+}
 
-    g->m.ter_set( act->placement, next_ter );
-
+bool activity_handlers::forage_results( map &ma, player *p, items_location &loc,
+                                        const tripoint &pos, bool pickup, bool messages )
+{
+    const int veggy_chance = rng( 1, 100 );
+    bool found_something = false;
+    bool inventory_space = true;
     // Survival gives a bigger boost, and Perception is leveled a bit.
     // Both survival and perception affect time to forage
 
     ///\EFFECT_PER slightly increases forage success chance
     ///\EFFECT_SURVIVAL increases forage success chance
     if( veggy_chance < p->get_skill_level( skill_survival ) * 3 + p->per_cur - 2 ) {
-        const auto dropped = g->m.put_items_from_loc( loc, p->pos(), calendar::turn );
-        for( const auto &it : dropped ) {
-            add_msg( m_good, _( "You found: %s!" ), it->tname() );
+        std::vector<item *> dropped = ma.put_items_from_loc( loc, pos, calendar::turn );
+        for( item *it : dropped ) {
+            if( messages ) {
+                add_msg( m_good, _( "You found: %s!" ), it->tname() );
```

Comment:
all these `add_msg()` need to be `add_msg_player_or_npc` and need to be gated on whether the player can see the NPC or else the player is going to get spammed by messages for NPCs harvesting bushes on the other side of the world.

Reply:
`add_msg_player_or_npc` already checks whether the player can see the NPC (or monster) and only displays the message if the player can see it.